webapi/src/main/java/org/sakaiproject/webapi/controllers/TasksController.java

```diff
@@ -67,27 +93,149 @@
+    public Object[] getSiteGroups(@PathVariable String siteId) throws IdUnusedException {
+        checkSakaiSession();
+
+        Site currentSite = siteService.getSite(siteId);
+        Map<String, String> groupsMap = new HashMap<>();
+        
+        Collection groups = currentSite.getGroups();
+        if (!groups.isEmpty()) {
+            groupsMap = (Map<String, String>) groups.stream().collect(Collectors.toMap(Group::getId, Group::getTitle));
+        }
+
+        return groupsMap.entrySet().toArray(); 
+    }
 
     @PostMapping(value = "/tasks", produces = MediaType.APPLICATION_JSON_VALUE)
     public UserTaskAdapterBean createTask(@RequestBody UserTaskAdapterBean taskTransfer) {
 
         checkSakaiSession();
-        return UserTaskAdapterBean.from(taskService.createSingleUserTask(taskTransfer));
+
+        UserTaskAdapterBean result = taskTransfer;
+        SimpleDateFormat sdf = new SimpleDateFormat("yyyyMMddHHmmssSSS");
+        String time = sdf.format(new Date());
+
+        Task task = new Task();
+        BeanUtils.copyProperties(taskTransfer, task);
+        task.setReference("/user/" + taskTransfer.getUserId() + "/" + time);
+        task.setSystem(false);
+        String assignationType = taskTransfer.getAssignationType();
+        if (AssignationType.site.name().equals(assignationType)) {
+            try {
+                Site site = siteService.getSite(taskTransfer.getSiteId());
+                Set<String> users = site.getUsersIsAllowed("section.role.student");
+                task = taskService.createTask(task, users, taskTransfer.getPriority());
+                result = UserTaskAdapterBean.from(taskService.createUserTask(task, taskTransfer));
+                taskService.assignTask(task, AssignationType.site, taskTransfer.getSiteId());
+                result.setTaskAssignedTo(getTaskAssignedDescription(task.getId(), site));
+                result.setSiteTitle(site.getTitle());
+            } catch (BeansException | IdUnusedException e) {
+                log.error(e.getMessage(), e);
+            }
+        } else if (AssignationType.group.name().equals(assignationType)) {
+            try {
+                Site site = siteService.getSite(taskTransfer.getSiteId());
+                Set<String> users = new HashSet();
+                List<String> groups = new ArrayList();
+                for (String groupId : taskTransfer.getSelectedGroups()) {
+                    groups.add(groupId);
+                    Group group = site.getGroup(groupId);
+                    Set<Member> members = group.getMembers();
+                    for (Iterator membersIter = members.iterator(); membersIter.hasNext();) {
+                        Member member = (Member) membersIter.next();
+                        users.add(member.getUserId());
+                    }
+                }
+                task = taskService.createTask(task, users, taskTransfer.getPriority());
```

Comment:
What happens when a user is removed from the group? Does their task get cleaned up?